subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```diff
@@ -538,3 +635,317 @@ int nrf_cloud_decode_data_endpoint(const struct nrf_cloud_data *input,
+		k_free(job->id);
+	}
+
+	if (job->host) {
+		k_free(job->host);
+	}
+
+	if (job->path) {
+		k_free(job->path);
+	}
+
+	memset(job, 0, sizeof(*job));
+	job->type	= NRF_CLOUD_FOTA_TYPE__INVALID;
+
+	return ret;
+}
+
+int nrf_cloud_parse_pgps_response(const char *const response,
+	struct nrf_cloud_pgps_result *const result)
+{
+	if (!response || !result ||
+	    !result->host || !result->host_sz ||
+	    !result->path || !result->path_sz) {
+		return -EINVAL;
+	}
+
+	char *host_ptr = NULL;
+	char *path_ptr = NULL;
+	int err = 0;
+	cJSON *rsp_obj = cJSON_Parse(response);
+
+	if (!rsp_obj) {
+		LOG_ERR("P-GPS response does not contain valid JSON");
+		err = -EBADMSG;
+		goto cleanup;
+	}
+
+	/* MQTT response is an array, REST is key/value map */
+	if (cJSON_IsArray(rsp_obj)) {
+		if (get_string_from_array(rsp_obj, PGPS_RCV_ARRAY_IDX_HOST, &host_ptr) ||
+		    get_string_from_array(rsp_obj, PGPS_RCV_ARRAY_IDX_PATH, &path_ptr)) {
+			LOG_ERR("Invalid P-GPS array response format");
+			err = -EFTYPE;
+			goto cleanup;
+		}
+	} else if (get_string_from_obj(rsp_obj, PGPS_RCV_REST_HOST, &host_ptr) ||
+		   get_string_from_obj(rsp_obj, PGPS_RCV_REST_PATH, &path_ptr)) {
+		LOG_ERR("Invalid P-GPS REST response format");
+		err = -EFTYPE;
+		goto cleanup;
+	}
+
+	if (!host_ptr || !path_ptr) {
+		err = -ENOSTR;
+		goto cleanup;
+	}
+
+	if ((result->host_sz <= strlen(host_ptr)) ||
+	    (result->path_sz <= strlen(path_ptr))) {
+		err = -ENOBUFS;
+		goto cleanup;
+	}
+
+	strncpy(result->host, host_ptr, result->host_sz);
+	LOG_DBG("host: %s", log_strdup(result->host));
+
+	strncpy(result->path, path_ptr, result->path_sz);
+	LOG_DBG("path: %s", log_strdup(result->path));
+
+cleanup:
+	if (rsp_obj) {
+		cJSON_Delete(rsp_obj);
+	}
+	return err;
+}
+
+int get_string_from_array(const cJSON * const array, const int index,
+			  char **string_out)
+{
+	__ASSERT_NO_MSG(string_out != NULL);
+
+	cJSON *item = cJSON_GetArrayItem(array, index);
+
+	if (!cJSON_IsString(item)) {
+		return -EINVAL;
+	}
+
+	*string_out = item->valuestring;
+
+	return 0;
+}
+
+int get_string_from_obj(const cJSON * const obj, const char *const key,
+			char **string_out)
+{
+	__ASSERT_NO_MSG(string_out != NULL);
+
+	cJSON *item = cJSON_GetObjectItem(obj, key);
+
+	if (!cJSON_IsString(item)) {
+		return -EINVAL;
+	}
+
+	*string_out = item->valuestring;
+
+	return 0;
+}
+
+char * const nrf_cloud_format_cell_pos_req_payload(struct lte_lc_cells_info const * const inf,
+	size_t inf_cnt)
+{
+	if (!inf || !inf_cnt) {
+		return NULL;
+	}
+
+	char * payload = NULL;
+	cJSON *lte_array;
+	cJSON *payload_obj = cJSON_CreateObject();
+
+	lte_array = cJSON_AddArrayToObject(payload_obj, NRF_CLOUD_CELL_POS_JSON_KEY_LTE);
+	if (!lte_array) {
+		goto cleanup;
+	}
+
+	for (size_t i = 0; i < inf_cnt; i++) {
+		struct lte_lc_cell const * const cur = &(inf + i)->current_cell;
+		cJSON *lte_obj = cJSON_CreateObject();
+
+		if (!lte_obj) {
+			goto cleanup;
+		}
+		cJSON_AddItemToArray(lte_array, lte_obj);
```

Comment:
Potential leak here for `lte_obj` if `cJSON_AddItemToArray()` fails, it will live on as a detached object

Reply:
multi-cell is still a work in progress... i just pushed what i had done before leaving yesterday 😅